### ISSUE TITLE
KPool, visually: finding 2K useful tokens without searching every token individually

### ISSUE THREAD

**Issue by raikonenfnu** [raikonenfnu/tokenspeed#67]
## The question

Suppose a model is reading a book with 128,000 highlighted passages.

For every new word it writes, it wants to retrieve roughly 2,000 relevant earlier tokens. The expensive approach is to ask, one by one:

> “How relevant is token 1? Token 2? Token 3? … token 128,000?”

KPool changes the first question. It asks about **small neighborhoods** before opening every house in them.

This issue builds an intuition for that mechanism, then connects the picture to the tensors and kernels an inference engineer sees.

---

### Dense attention searches the entire past

In ordinary causal self-attention, every new query compares itself with every visible key in the history:

```text
new query q
    │
    ├── compare with k₀
    ├── compare with k₁
    ├── compare with k₂
    ├── ...
    └── compare with kₙ₋₁
```

It then turns those similarities into probabilities and forms a weighted mixture of all values:

```math
A(q,K,V) = \mathrm{softmax}\!\left(\frac{qK^\mathsf{T}}{\sqrt d}\right)V
```

This is wonderfully direct: the query can inspect every earlier token. But during decoding, each layer must repeatedly read a KV history whose width grows with context length. At 128K tokens, even generating one new token means traversing an enormous memory shelf. The GPU often spends more effort moving old keys and values than doing arithmetic on them.

### Sparse attention asks a retrieval question first

Sparse attention separates the operation into two stages:

```text
                         cheap search
query + long history ───────────────────▶ promising token positions
                                                  │
                                                  ▼
                                      expensive, precise attention
                                                  │
                                                  ▼
                                               output
```

Rather than run full attention over all `N` historical tokens, a lightweight **indexer** scores the history and selects only `K` candidates, where `K ≪ N`. Full attention then reads the original keys and values for those candidates.

For a fixed selected width of roughly 2,048 tokens:

```text
128K-token context

Dense attention:   inspect all 131,072 original KV entries
Sparse attention:  search a light index, then inspect ~2,048 original KV entries
```

The computation is sparse, but the information does not have to be permanently deleted: a different query may select a completely different subset of history.

### DeepSeek Sparse Attention: a learned lightning indexer

DeepSeek Sparse Attention (DSA) uses a learned, multi-head indexer. It projects each new token into small index queries and each historical token into a small index key. For candidate token `j`, each index head produces a similarity; ReLU keeps positive evidence, and learned query-dependent weights combine the heads:

```math
S_j = \sum_h w_h\,\mathrm{ReLU}\!\left(\gamma q_h^\mathsf{T}k_j^{\mathrm{index}}\right)
```

The top-scoring positions become the sparse attention set. This index is cheaper than reading the full MLA keys and values, and training teaches it which tokens are likely to matter to the real attention operation.

A helpful analogy is a library:

- **Dense attention** pulls every book from the shelves and reads the relevant pages.
- **Sparse attention** searches a compact catalog, then pulls only the promising books.
- **The indexer** is the catalog and ranking system.
- **Full sparse MLA** is the careful reading performed after retrieval.

### The remaining problem: the catalog can become large too

Sparse attention removes most work from the final attention kernel, but a token-level indexer still scores one index key for every historical token:

```text
final attention width:     N  →  K
index search width:        N  →  N
```

At very long context, scanning that catalog is itself substantial. If every four raw tokens can be represented by one useful search card, the catalog becomes four times shorter:

```text
final attention width:     N  →  about 2,048
index search width:        N  →  N / 4
```

This is where KPool connects.

### KPool is hierarchical sparse retrieval

KPool adds a coarse level above token-level sparse attention:

```text
original KV tokens
      │
      ├── group neighboring index keys into learned four-token pools
      │
      ├── score the shorter pool index
      │
      ├── retain the best 512 pools
      │
      ├── expand each pool back to its four original positions
      │
      └── run precise sparse MLA on ~2,048 original KV entries
```

So the relationship can be summarized as:

> **Sparse attention makes attention selective; KPool makes the selection process itself more selective and compact.**

DSA without KPool resembles searching every entry in a catalog and retrieving the best 2,048 books. KPool resembles first searching 32,768 shelf labels, choosing 512 shelves, and then retrieving the four books on every chosen shelf.

That extra hierarchy introduces a tradeoff. It reduces index memory traffic and scoring work, but selection happens at neighborhood granularity: if one token is important, its three pool neighbors come along too. The learned compressor must therefore make an important token cause its whole pool to rank highly enough. Final attention remains precise because it still sees the four original KV entries and can assign nearly all attention probability to just one of them.

---

## Scene 1: turn a long row of tokens into tiles

Imagine the KV history as colored dots:

```text
raw token history

● ● ● ●   ● ● ● ●   ● ● ● ●   ● ● ● ●   ...
└── 4 ──┘  └── 4 ──┘  └── 4 ──┘  └── 4 ──┘
 pool 0     pool 1     pool 2     pool 3
```

For the production geometry discussed here, one pool represents four consecutive tokens.

Each group produces a small **index card**: a single 128-dimensional vector that summarizes where the four keys point in semantic space.

```text
(k₀, k₁, k₂, k₃)  ── learned compression ──▶  p₀
(k₄, k₅, k₆, k₇)  ── learned compression ──▶  p₁
```

The card is compactly stored in scaled FP8. The original four KV entries are not discarded; they remain in the ordinary KV cache.

This creates two parallel memories:

```text
                    ┌─ compact KPool index: one card per 4 tokens
history ────────────┤
                    └─ original KV cache: every individual token
```

The first memory answers **where should I look?**  
The second answers **what should I attend to?**

That separation is the key idea.

---

## Scene 2: the pool is not an ordinary average

A plain mean would blur together four unrelated keys. Instead, each coordinate of the pooled key gets a learned vote.

For pool slot `s ∈ {0, 1, 2, 3}` and vector coordinate `d`:

```math
\alpha_{s,d}
= \mathrm{softmax}_s\big(g_{s,d} + a_{s,d}\big)
```

```math
p_d = \sum_{s=0}^{3} \alpha_{s,d} k_{s,d}
```

- `k` is the index key generated for each token.
- `g` is a content-dependent gate generated by the model.
- `a` is a learned within-pool positional bias.
- The softmax runs across the four positions separately for each coordinate.

A useful picture is four translucent images laid on top of one another. For one feature, token 0 may be bright; for another feature, token 3 may be bright. KPool can preserve different evidence from different positions in the same pooled vector.

Before FP8 storage, the implementation applies an orthogonal Hadamard rotation. Intuitively, it spreads unusually large coordinates across the vector so quantization is less dominated by a few outliers.

---

## Scene 3: let the new query shine a flashlight

For a new query, the indexer creates multiple query heads. Each head compares the query with every pooled card:

```math
r_{h,j} = q_h^\top p_j
```

The scores are scaled, passed through a ReLU, and combined using learned query-dependent head weights:

```math
R_j = \sum_h w_h\,\mathrm{ReLU}(\gamma r_{h,j})
```

Now imagine every pool as a tile on a very long floor. The query shines 32 differently colored flashlights on the tiles; the learned head weights decide how much each color matters. A top-k operation keeps the 512 brightest tiles.

```text
all pooled cards                         selected cards

□ □ ■ □ □ □ ■ □ □ ■ □ □ ...  ──────▶   ■ ■ ■ ...  (top 512)
```

Because each selected tile represents four original tokens:

```text
512 selected pools × 4 tokens/pool = 2,048 raw-token candidates
```

Those pool IDs are expanded back into physical slots in the original FlatKV cache. Sparse MLA then performs real attention over the original keys and values at those slots.

```text
query
  │
  ├─ score compact pool cards
  │
  ├─ choose 512 pools
  │
  ├─ expand pool IDs → about 2,048 raw KV positions
  │
  └─ run attention on the original KV entries
```

KPool therefore does **not** replace attention values with blurry pooled values. Pooling is the map, not the territory.

---

## Scene 4: the unfinished tile

During autoregressive decoding, the history length is rarely divisible by four.

```text
completed pools                live tail

● ● ● ●   ● ● ● ●   ● ● ● ●   ● ● · ·
                                └ 2 new tokens
```

The final one to three tokens cannot form a completed pool yet. They live in a small request-local tail buffer and are always appended to the selected candidates, subject to causality.

So the selected width is:

```math
4 \times 512 + \text{tail}, \qquad \text{tail} \in \{0,1,2,3\}
```

or at most 2,051 token slots. When the fourth token arrives, the tile closes: its four raw index keys are compressed, quantized, and written as one new pooled card.

This tail rule is easy to overlook, but it prevents the newest context from becoming temporarily invisible just because its pool is incomplete.

---

## Why this helps

Let the raw context length be `N`, pool size be `P=4`, and selected pool count be `K=512`.

Without pooling, an indexer scores `N` token-level keys. With KPool it scores approximately `N/P` compact keys, then full attention reads only about `P×K = 2,048` original KV entries.

```text
index search width:       N       → N / 4
full-attention KV width:  N       → ≈ 2,048
```

The first reduction is approximately 4× in index rows, not automatically 4× end-to-end speed. Runtime also contains top-k selection, cache addressing, candidate expansion, attention, communication, and the rest of the model. At shorter contexts, fixed kernel-launch and selection costs can dominate; at long contexts, reading and scoring the index becomes increasingly important.

KPool is best viewed as **learned coarse-to-fine retrieval**:

1. compress locally;
2. rank globally at pool granularity;
3. reopen the winning neighborhoods;
4. attend precisely to their original tokens.

---

## A geometric picture

A single pooled vector cannot reproduce all four keys. That is not its job. It only needs to make a relevant neighborhood rank highly enough to survive the first cut.

Think of looking for mountains from an airplane:

- The pooled card is the silhouette of a four-token neighborhood.
- Top-k chooses the silhouettes that look promising.
- Sparse attention lands in those neighborhoods and examines the original terrain.

This also reveals the core accuracy tradeoff. If four keys point in very different directions, their one-card silhouette may hide a key that an unpooled token indexer would have selected. The learned gates and positional bias exist to make that compression useful, and training must teach the index to preserve retrieval recall.

---

## Systems view

A practical implementation has four distinct jobs:

| Stage | What happens | Main systems concern |
|---|---|---|
| Pool write | Four raw index keys become one learned pooled key | fusion, FP8 quantization, cache writes |
| Score | Query heads dot-product against pooled FP8 keys | memory bandwidth and matrix throughput |
| Select | Head scores are reduced and top 512 pools are chosen | selection/sort cost, temporary workspace |
| Expand | 512 pool IDs become ~2,048 FlatKV slots | paging, causality, graph-safe output shapes |
| Attend | Sparse MLA reads original KV at those slots | attention-kernel efficiency |

Prefill adds ragged causal geometry: every query token may see a different number of completed pools. Decode is structurally simpler but latency-sensitive, so fusing scoring, head reduction, and top-k can matter more than maximizing any one kernel's throughput.

---

### “KPool means pooled attention”

No. The pooled keys are an index used for candidate selection. Final sparse attention consumes the original KV entries.

### “It is average pooling”

No. Compression is learned, position-aware, and coordinate-wise softmax weighted.

### “Selecting 512 pools means attending to 512 tokens”

No. With pool size four, it expands to 2,048 original token slots, plus up to three live-tail tokens.

### “The newest tokens wait until their pool is complete”

No. The visible incomplete tail is explicitly appended.

### “A 4× smaller index guarantees a 4× faster model”

No. It reduces one important component. The realized gain depends on context length, batch size, kernel fusion, top-k cost, memory traffic, attention cost, and overlap.

---

## Questions worth measuring

This mental model suggests concrete optimization questions:

- At what context length does pooled-index scoring become bandwidth-bound?
- Is top-k selection or score computation dominant at batch 1?
- Can score, ReLU, head reduction, and selection share a kernel or remain on-chip?
- Does the FP8 index reduce memory traffic enough to offset dequantization?
- How much work is wasted scoring graph-padded or causally invisible pools?
- Can completed-pool writes be hidden behind the rest of decode?
- What is pool-level recall relative to an equivalent token-level index?
- Would a different pool size improve the quality/performance frontier?

The right benchmark should therefore report the stages separately rather than labeling all of them “sparse attention.”

---

## Further reading

- [DeepSeek-V3.2 technical report](https://huggingface.co/deepseek-ai/DeepSeek-V3.2/resolve/main/assets/paper.pdf) — the learned sparse-attention/indexer foundation.
- [HISA: Hierarchical Index Sparse Attention](https://arxiv.org/abs/2603.28458) — closely related coarse-to-fine pooled indexing.
- [GLM-5.3-Flash architecture notes](https://sebastianraschka.com/blog/2026/glm-5-3-flash-architecture-notes.html) — accessible model-level context.
- [DSA implementation walkthrough](https://github.com/rasbt/LLMs-from-scratch/blob/main/ch04/09_dsa/README.md) — code-oriented explanation of the underlying sparse-attention idea.

---

## One-sentence summary

**KPool folds every four-token neighborhood into a learned search card, uses those cards to find 512 promising neighborhoods, then unfolds them and performs precise attention over the original ~2,048 KV tokens.**
